### PR TITLE
Added TypeScript 2 example project

### PR DESCRIPTION
**Pull request by mceachen**
I've added a partial typings for Objection and an example project that consumes the typings in the PR to address #102. I've got a solid 3 weeks of experience with TypeScript at this point, so temper expectations accordingly.

This PR compiles and runs successfully on node v6.8.0 on linux and win10.

As this typing is self-contained in the examples directory, you can merge this PR and not adversely impact anyone except others that are interested in pushing forward high-quality typings for Objection.

Remaining tasks after accepting this PR:

- [ ] Submit this to [DefinitelyTyped ](http://definitelytyped.org/guides/contributing.html). 
- [ ] Remove the typings directory and add the new types to `examples/express-ts/package.json`.

### CODE REVIEW

**Comment by drew-r** [comment]
Thanks @mceachen 
How is `bindKnex()` intended to be used with these type defs?

**Comment by mceachen** [comment]
@drew-r that's a mistake--I misread the API guideline, and it should return a typeof Model (or in this case, `typeof T`, but that [doesn't look like it's supported](https://github.com/Microsoft/TypeScript/issues/204). 

I'll look into it a more tomorrow.

**Comment by drew-r** [comment]
I fixed it and a few other things, whats the best way to get them merged?

**Comment by mceachen** [comment]
@drew-r where's your diff or branch?

**Comment by mceachen** [comment]
I believe subclasses are properly handed in 0bcad79 now, so `Model` types are ready. 

`QueryBuilder` is still a WIP. I'll submit another diff finishing that work soon.

File: examples/express-ts/typings/objection/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,692 @@
+// Type definitions for objection v0.6.1
+// Project: Objection.js <http://vincit.github.io/objection.js/>
+// Definitions by: [~YOUR NAME~] <[~A URL FOR YOU~]>
```
Comment (elhigu) [inline examples/express-ts/typings/objection/index.d.ts]: Your name and url missing

File: examples/express-ts/typings/objection/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+  type ModelsOrObjects = Model | Object | Model[] | Object[];
+
+  type RelationOptions = { alias: boolean | string };
+
+  interface JoinRelation {
+    (relationName: string, opt?: RelationOptions): QueryBuilder;
+  }
+
+  interface WhereJson {
+    (fieldExpression: FieldExpression, jsonObjectOrFieldExpression: any): QueryBuilder;
```
Comment (elhigu) [inline examples/express-ts/typings/objection/index.d.ts]: Why here `jsonObjectOrFieldExpression` is `any` and not like below?

File: examples/express-ts/typings/objection/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+
+    deleteById(idOrIds: IdOrIds): QueryBuilder;
+
+    relate(ids: IdOrIds | ModelsOrObjects): QueryBuilder;
+    unrelate(): QueryBuilder;
+
+    forUpdate(): QueryBuilder;
+    forShare(): QueryBuilder;
+
+    // TODO: fromJS does not exist in current knex documentation: http://knexjs.org/#Builder-fromJS
```
Comment (elhigu) [inline examples/express-ts/typings/objection/index.d.ts]: I think from JS was removed from knex at some point... I had never used it either.

**Comment by drew-r** [comment]
@mceachen I wonder is there a way to have QueryBuilder return Promises that are typed to the model?
I played around with a `QueryBuilder<T>` and a `query(this: T): QueryBuilder<T>` but it kind of breaks down when you have to start subclassing models for example with `bindKnex`, as `this: T` is a `typeof Model` rather than `Model`. Any ideas?

**Comment by mceachen** [comment]
@drew-r I played with that a bit, but because you can make queries that
return partial models,  model unions, and even arbitrary columns, that
discouraged me from pushing forward with that idea.

If there was a query builder with a much simpler interface, I think that'd
be great. The current query builder would presumably then subclass and
extend this strongly typed query builder.

@elhigu is there anything else I can do to get this PR merged? Thanks for
your prior review!

**Comment by danieloprado** [comment]
A simple solution is type the return function:

```typescript
export async function list(churchId: number): Promise<Informative[]> {
  return await Informative.query().where({ churchId });
}

export async function insert(informative: IInformative): Promise<Informative> {
  return await Informative.query().insert(informative);
}
```

Works like a charm ;)

**Comment by drew-r** [comment]
thanks both.
@mceachen that makes perfect sense. 
While I have your ear so to speak, do you see an elegant way to expose the static properties defined on Model classes when using `bindKnex()`?

**Comment by elhigu** [comment]
@mceachen all good, let us know about Definedlytyped stuff how it goes 👍

**Comment by elhigu** [comment]
Aand thanks! I look forward to test these in our own project which we've been writing with typescript :)

**Comment by drew-r** [comment]
Also there are a few edits that need to be made... `query()` needs an
optional transaction parameter, `transaction` itself needs to be exported
additionally as a namespace with the static `Transaction.start()` relocated
and exported within.
fromJson options are optional also I believe.

Sorry I would put a patch together but restricted to mobile right now

**Comment by mceachen** [comment]
Ah OK, I'll make a new pr with those changes, and link to the definitely
typed pr, too.

**Comment by mceachen** [comment]
@drew-r just created #246 to address your comments.

**Comment by Amerzel** [comment]
I would like to see these submitted to Definitely Typed to make them easier to use. Is that still on the radar @mceachen?

Thanks for doing the work and adding these.

**Comment by mceachen** [comment]
I had one more set of changes before I wanted to submit to DT, which are here: #254. I'll reference the DT PR here when I post it.

**Comment by mceachen** [comment]
@Amerzel In reading the best-practices from DefinitelyTyped, they recommend the library self-host typings. See #255 which moves the types from the example and into the objection module proper.

**Comment by Amerzel** [comment]
@mceachen,

That makes sense. There's pluses and minuses to doing it that way. It makes it really easy to use the types but adds a bit more work for the library maintainers. They would be expected to maintain and update the typings as new features are added to the library. If the maintainers don't use TypeScript having the typings self hosted adds some additional work that they may not care about or want to maintain.

James

**Comment by elhigu** [comment]
@Amerzel yep, we've discussed about it a bit in gitter chat. The same amount of maintenance work is in both options: Someone has to update typings when new APIs are released, if there is no one who is committed to keep them updated for releases, then we can just drop the support. 

Good thing is that if we keep typings in the same repo, it is easy to keep versions synched that correct version of typings comes directly with the certain library version (I suppose that definitely typed has also mechanism for it, but to me keeping everything in the same repo feels easier).

**Comment by Amerzel** [comment]
Sounds great. I'm glad it was discussed. Keeping everything in the same repo is definitely easier :)

**Pull request merged**